go/libraries/doltcore/env/actions/infer_schema.go

```diff
@@ -0,0 +1,419 @@
+	}
+
+	orderedPKCols := make([]schema.Column, len(pkCols))
+	for _, col := range pkCols {
+		idx, ok := inf.pkColToIdx[col.Name]
+
+		if !ok {
+			return nil, errors.New("could not find key column")
+		}
+
+		orderedPKCols[idx] = col
+	}
+
+	colColl, err := schema.NewColCollection(cols...)
+
+	if err != nil {
+		return nil, err
+	}
+
+	pkColColl, err := schema.NewColCollection(orderedPKCols...)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return schema.SchemaFromPKAndNonPKCols(pkColColl, colColl)
+}
+
+func nextTag(tag uint64, cols *schema.ColCollection) uint64 {
+	for {
+		_, ok := cols.GetByTag(tag)
+
+		if !ok {
+			return tag
+		}
+
+		tag++
+	}
+}
+
+func typeCountsToKind(name string, typeToCount map[types.NomsKind]int, hasNegatives bool) (types.NomsKind, bool) {
+	var nullable bool
+	kind := types.NullKind
+
+	for t := range typeToCount {
+		if t == types.NullKind {
+			nullable = true
+			continue
+		} else if kind == types.NullKind {
+			kind = t
+		}
+
+		if kind == t {
+			continue
+		}
+
+		switch kind {
+		case types.StringKind:
+			if nullable {
+				return types.StringKind, true
+			}
+
+		case types.UUIDKind:
+			//cli.PrintErrln(color.YellowString("warning: column %s has a mix of uuids and non uuid strings.", name))
+			kind = types.StringKind
+
+		case types.BoolKind:
+			kind = types.StringKind
+
+		case types.IntKind:
+			if t == types.FloatKind {
+				kind = types.FloatKind
+			} else if t == types.UintKind {
+				if !hasNegatives {
+					kind = types.UintKind
+				} else {
+					//cli.PrintErrln(color.YellowString("warning: %s has values larger than a 64 bit signed integer can hold, and negative numbers.  This will be interpreted as a string.", name))
+					kind = types.StringKind
+				}
+			} else {
+				kind = types.StringKind
+			}
+
+		case types.UintKind:
+			if t == types.IntKind {
+				if hasNegatives {
+					//cli.PrintErrln(color.YellowString("warning: %s has values larger than a 64 bit signed integer can hold, and negative numbers.  This will be interpreted as a string.", name))
+					kind = types.StringKind
+				}
+			} else {
+				kind = types.StringKind
+			}
+
+		case types.FloatKind:
+			if t != types.IntKind {
+				kind = types.StringKind
+			}
+		}
+	}
+
+	if kind == types.NullKind {
+		kind = types.StringKind
+	}
+
+	return kind, nullable
+}
+
+func (inf *inferrer) sinkRow(p *pipeline.Pipeline, ch <-chan pipeline.RowWithProps, badRowChan chan<- *pipeline.TransformRowFailure) {
+	for r := range ch {
+		i := 0
+		_, _ = r.Row.IterSchema(inf.sch, func(tag uint64, val types.Value) (stop bool, err error) {
+			defer func() {
+				i++
+			}()
+
+			if val == nil {
+				count, _ := inf.colType[i][types.NullKind]
+				inf.colType[i][types.NullKind] = count + 1
+				return false, nil
+			}
+
+			strVal := string(val.(types.String))
+			kind, hasNegs := leastPermissiveKind(strVal, inf.impArgs.FloatThreshold)
+
+			if hasNegs {
+				inf.negatives[i] = true
+			}
+
+			count, _ := inf.colType[i][kind]
```

Comment:
Does inf.colType[i][kind] += 1 not work?

Reply:
Syntactically it's fine but believe it or not go-lint would complain about it:
![image](https://user-images.githubusercontent.com/871117/67128547-d6b16980-f1b0-11e9-904a-bf44542b2240.png)

...I think inf.colType[i][kind]++ is fine, though kinda gross.

Reply:
you actually have to have a 0 value in the map to use these don't you?  can you increment an int in a map if there is no value for the key?

Reply:
just tested.  Answer is no.  I will update.